parasol/query.py

```diff
@@ -0,0 +1,241 @@
+"""
+Object-oriented approach to Solr searching and filtering modeled
+on :class:`django.models.queryset.QuerySet`.  Supports iteration,
+slicing, counting, and boolean check to see if a search has results.
+
+Filter, search and sort methods return a new queryset, and can be
+chained. For example::
+
+    SolrQuerySet().filter(item_type='person') \
+                  .search(name='hem*') \
+                  .order_by('sort_name') \
+
+
+"""
+
+from typing import Any, Optional
+
+try:
+    from parasol.solr.django import SolrClient
+except ImportError:
+    # FIXME: SolrQuerySet doesn't currently work with non-django
+    # solr client because it would need a way to get connection settings
+    from parasol.solr import SolrClient
+
+
+class SolrQuerySet:
+    """A Solr queryset object that allows for object oriented
+    searching and filtering of Solr results. Allows search results
+    to be paginated by django paginator."""
+
+    _result_cache = None
+    start = 0
+    stop = None
+    sort_options = []
+    search_qs = []
+    filter_qs = []
+
+    #: by default, combine search queries with AND
+    default_search_operator = 'AND'
+
+    def __init__(self, solr: Optional[SolrClient] = None):
+        self.solr = solr or SolrClient()
+        # convert search operator into form needed for combining queries
+        self._search_op = ' %s ' % self.default_search_operator
+
+    def get_results(self, **kwargs):
+        """
+        Query Solr and get the results for the current query and filter
+        options. Populates result cache and returns the documents portion
+        of the reponse.
+
+        Returns:
+            Solr response documents as a list of dictionaries.
+        """
+        query_opts = self.query_opts()
+        query_opts.update(**kwargs)
+        # TODO: what do we do about the fact that Solr defaults
+        # to 10 rows?
+
+        # NOTE: django templates choke on AttrDict because it is
+        # callable; using dictionary response instead
+        self._result_cache = self.solr.query(wrap=False, **query_opts)
+        return self._result_cache['docs']
+
+    def query_opts(self):
```

Comment:
Not sure if we're doing type annotations, but this would be a good candidate for one of the more complicate ones for return type `-> Dict[str, str]`